internal/servers/common/client_ip.go

```diff
@@ -0,0 +1,134 @@
+package common
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"strings"
+	"sync/atomic"
+
+	"github.com/hashicorp/boundary/internal/errors"
+)
+
+const (
+	RealIpHeader        = "X-Real-Ip"
+	XForwardedForHeader = "X-Forwarded-For"
+)
+
+var privateNets atomic.Value
+
+// ClientIpFromRequest will determine if the client IP of the http request using
+// the provide set of private networks. See InitPrivateNetworks(...) and
+// PrivateNetworks(...) for building the list of private networks.
+func ClientIpFromRequest(ctx context.Context, privateNetworks []*net.IPNet, r *http.Request) (string, error) {
+	const op = "ClientIpFromRequest"
+	if privateNetworks == nil {
+		return "", errors.New(ctx, errors.InvalidParameter, op, "missing list of private networks")
+	}
+	if r == nil {
+		return "", errors.New(ctx, errors.InvalidParameter, op, "missing http request")
+	}
+
+	forwardedFor := r.Header.Get(XForwardedForHeader)
+	realIp := r.Header.Get(RealIpHeader)
+
+	// no headers, so we'll use the remote addr
+	if forwardedFor == "" && realIp == "" {
+		if strings.ContainsRune(r.RemoteAddr, ':') {
+			if ip, _, err := net.SplitHostPort(r.RemoteAddr); err != nil {
+				return "", errors.Wrap(ctx, err, op)
+			} else {
+				return ip, nil
+			}
+		}
+		return r.RemoteAddr, nil
+	}
+
+	// next up, we'll see if there's any public addrs in the X-Forwarded-For
+	// header
+	for _, a := range strings.Split(forwardedFor, ",") {
+		a = strings.TrimSpace(a)
+		isPrivate, err := isPrivateAddr(ctx, privateNetworks, a)
+		if !isPrivate && err == nil {
+			return a, nil
+		}
+	}
+
+	// finally, just fallback to the contexts of the X-Real-Ip header (which
+	// could be empty)
+	return realIp, nil
```

Comment:
Would we really want to fall back to an empty `"X-Real-Ip"`?  Wouldn't it be better to fall back to `r.RemoteAddr`
Also if we have `X-Real-Ip` wouldn't that be better than `X-Forwarded-For`?

Seems logic here should be check for `X-Real-Ip` if empty check for `X-Forwarded-For` that is private, if none then fall back to r.RemoteAddr

Reply:
ty!